[pipelining] Add grad test for interleaved schedules

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #127066
* __->__ #126931
* #127136

Added `test_grad_with_manual_interleaved`:
- Model: `MultiMLP`
- Tested schedules: Interleaved1F1B, LoopedBFS
- Two stages per rank
```
Rank 0 stages: [0, 2]
Rank 1 stages: [1, 3]
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

kwen2501 committed:
[WIP][pipelining] Add grad test for interleaved schedules

[ghstack-poisoned]

kwen2501 committed:
Update on "[WIP][pipelining] Add grad test for interleaved schedules"


```
Traceback (most recent call last):
  File "/data/users/kw2501/pytorch/torch/testing/_internal/common_utils.py", line 2756, in wrapper
    method(*args, **kwargs)
  File "/data/users/kw2501/pytorch/torch/testing/_internal/common_utils.py", line 443, in instantiated_test
    test(self, **param_kwargs)
  File "/data/users/kw2501/pytorch/test/distributed/pipelining/test_schedule.py", line 316, in test_grad_with_manual_interleaved
    out = schedule.step(target=target, losses=losses)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/kw2501/pytorch/torch/distributed/pipelining/PipelineSchedule.py", line 578, in step
    self._step_microbatches(args_split, kwargs_split, targets_split, losses)
  File "/data/users/kw2501/pytorch/torch/distributed/pipelining/PipelineSchedule.py", line 820, in _step_microbatches
    ops.extend(bwd_stage.get_bwd_send_ops())
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/kw2501/pytorch/torch/distributed/pipelining/PipelineStage.py", line 339, in get_bwd_send_ops
    raise RuntimeError(
RuntimeError: [1] for chunk 0 has gradients None and is expecting to send gradients to stage 0

To execute this test, run the following from the base repo dir:
     python test/distributed/pipelining/test_schedule.py -k ScheduleTest.test_grad_with_manual_interleaved_ScheduleClass0
```

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wconstab commented on test/distributed/pipelining/test_schedule.py:
```diff
@@ -247,6 +251,96 @@ def test_grad_with_manual(self, ScheduleClass):
+        with torch.no_grad():
+            y = ref_mod(x)
+            # Add a small perturbation
+            target = y + torch.randn(batch_size, d_hid, device=self.device)
```
nice way of doing the grad test.

wconstab replied:
except, how did you determine the perturbation is 'small'? (is the default of randn small relative to the norm of the model's output?)

wconstab reviewed (approved):
nice! thanks for putting this in.

kwen2501 commented:
@pytorchbot merge

kwen2501 commented:
@pytorchbot merge -f "the pull or windows failure does not seem related"

clee2000 commented:
@pytorchbot revert -m "newly added test fails distributed/pipelining/test_schedule.py::ScheduleTest::test_grad_with_manual_interleaved_ScheduleClass0 https://hud.pytorch.org/pytorch/pytorch/commit/abf6d4e6bc1a9a0e08bfc2204560ca7858fa90cd https://github.com/pytorch/pytorch/actions/runs/9214413308/job/25352507591, pull workflow failed on startup on PR, so no distributed tests ran at all" -c nosignal

kwen2501 committed:
Update on "[pipelining] Add grad test for interleaved schedules"


Added `test_grad_with_manual_interleaved`:
- Model: `MultiMLP`
- Tested schedules: Interleaved1F1B, LoopedBFS
- Two stages per rank
```
Rank 0 stages: [0, 2]
Rank 1 stages: [1, 3]
```

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge
